[wasm] Some cleanup of targets, especially around profiler use

radical opened the pull request.

radical committed:
[wasm] samples: makefile cleanup

radical committed:
[wasm] Mark FileWrites as Output

radical committed:
[wasm] Move driver-gen.c generation for and profile, !aot case to separate target

radical committed:
[wasm] Add $(WasmProfilers) for aot compiler

radical committed:
[wasm] update public properties/items list in WasmApp.targets

radical committed:
wasmapp.targets - some more cleanup - differentiate internal properties

radical committed:
[wasm] fix MonoAOTCompiler build error

vargaz commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -6,14 +6,35 @@
+      - $(WasmNativeStrip)   - Whenever to strip the native executable. Defaults to true.
+      - $(RunAOTCompilation) - Defaults to false.
+      - $(EnableProfiler)    - Defaults to false.
+      - $(WasmDebugLevel)
```
It would be simpler to just have WasmProfilers, and set EnableProfilers if its non-empty.

vargaz commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -6,14 +6,35 @@
+      - $(AOTMode)           - Defaults to `LLVMOnly`
+      - $(AotProfilePath)    - profile data file to be used for profile-guided optimization
+      - $(BuildAOTProfiled)  - Defaults to false.
+      - $(WasmInvariantGlobalization)
```
Same if AotProfilePath is set, build with profiled.

radical committed:
[wasm] WasmApp.targets - Add `WasmExtraConfig`, to add to `mono-config.js`

`WasmAppBuilder.EnableProfiler` was added to add a special entry to
`mono-config.js` (`config.enable_profiler=true`), for the
`browser-profile` sample. Instead of doing that, add a general way to
add extra json elements to the config file.

This can be done through `WasmAppBuilder.ExtraConfig` set via
`@(WasmExtraConfig)`. Examples:

```xml
    <WasmExtraConfig Include="enable_profiler" Value="true" />

    <!-- Value attribute can have a number, bool, quoted string, or json string -->

    <WasmExtraConfig Include="json" Value="{ &quot;abc&quot;: 4 }" />
    <WasmExtraConfig Include="string_val" Value="&quot;abc&quot;" />
    <WasmExtraConfig Include="string_with_json" Value="&quot;{ &quot;abc&quot;: 4 }&quot;" />
```

This would add the following to the config file:

```json
  "enable_profiler": true,
  "json": {
    "abc": 4
  },
  "string_val": "abc",
  "string_with_json": "{ \"abc\": 4 }"
```

radical committed:
[wasm] Remove BuildAOTProfiled, and instead use AOTProfilePath only

radical committed:
[wasm] Fix typo

radical committed:
[wasm] MonoAOTCompiler: fix NRE

radical committed:
[wasm] WasmAppBuilder.ExtraConfig: add comments

radical commented on src/mono/wasm/build/WasmApp.targets:
```diff
@@ -195,6 +214,25 @@ EMSCRIPTEN_KEEPALIVE void mono_wasm_load_profiler_aot (const char *desc) { mono_
+  <Target Name="_GenerateDriverGenC" Condition="'$(RunAOTCompilation)' != 'true' and '$(WasmProfilers)' != ''">
+    <PropertyGroup>
+      <EmccFlags>$(EmccFlags) -DDRIVER_GEN=1</EmccFlags>
+      <InitAotProfilerCmd>
+void mono_profiler_init_aot (const char *desc)%3B
+EMSCRIPTEN_KEEPALIVE void mono_wasm_load_profiler_aot (const char *desc) { mono_profiler_init_aot (desc)%3B }
+      </InitAotProfilerCmd>
```
@vargaz given `"'$(RunAOTCompilation)' != 'true' and '$(WasmProfilers)' != '`, could `WasmProfilers` have more than just `aot`? If so, then I should be generating these 2 lines for all of them.

vargaz replied:
Yes, but right now we probably only use aot.

radical committed:
[wasm] WasmAppBuilder - fix json escaping
